File: icsm31-projeto/server-compiled/signal_gain.go
// Package main — Ganho de sinal aplicado ao vetor g, sem gonum.
// Enunciado: "Calculo do ganho de sinal (gamma)".
package main

// ApplySignalGain aplica gamma_l = 100 + (1/20)*l*sqrt(l) a cada amostra l
// do sinal, in-place. Espera-se g organizado em ordem column-major: para cada
// sensor c em [0,N), as S amostras consecutivas correspondem a l = 1..S.
// sqrt e implementado no proprio projeto (ver linalg.go), sem o pacote math.
func ApplySignalGain(g []float64, S, N int) {
	n := len(g)
	if n != S*N {
		// sinal 1D de tamanho S (ou menor): aplica gamma amostra a amostra
		for l := 0; l < n && l < S; l++ {
			ll := float64(l + 1)
			gamma := 100.0 + (1.0/20.0)*ll*Sqrt(ll)
			g[l] *= gamma
		}
		return
	}

	for c := 0; c < N; c++ {
		base := c * S
		for l := 0; l < S; l++ {
			ll := float64(l + 1)
			gamma := 100.0 + (1.0/20.0)*ll*Sqrt(ll)
			g[base+l] *= gamma
		}
	}
}

File: icsm31-projeto/server-interpreted/signal_gain.py
"""
Ganho de sinal aplicado ao vetor g antes da reconstrucao.
Enunciado: "Calculo do ganho de sinal (gamma)".

Formula:
    for c = 1 .. N:
        for l = 1 .. S:
            gamma_l = 100 + (1/20) * l * sqrt(l)
            g[l, c] = g[l, c] * gamma_l

O vetor g chega achatado em ordem coluna-a-coluna (Fortran): o indice plano
k mapeia para a amostra l = k % S e o sensor c = k // S. Assim basta
multiplicar cada posicao pelo gamma da sua amostra l. sqrt e implementado no
proprio projeto (modulo `linalg`), sem a lib math.
"""

from __future__ import annotations

from typing import List, Sequence

from linalg import sqrt


def apply_signal_gain(g: Sequence[float], S: int, N: int) -> List[float]:
    """Aplica o ganho gamma_l a cada amostra l do sinal.

    Args:
        g: vetor de sinal achatado (tamanho S*N, ordem coluna-a-coluna) ou de
           tamanho S (um unico sensor).
        S: numero de amostras por sensor.
        N: numero de sensores.

    Returns:
        nova lista com o ganho aplicado (mesmo tamanho do input).
    """
    # gamma_l = 100 + (1/20) * l * sqrt(l), para l = 1..S
    gamma = [100.0 + (1.0 / 20.0) * l * sqrt(l) for l in range(1, S + 1)]

    n = len(g)
    if n == S * N:
        # sinal completo (S*N): posicao k pertence a amostra l = k % S
        return [g[k] * gamma[k % S] for k in range(n)]

    # sinal 1D de tamanho S (ou menor): aplica gamma amostra a amostra
    return [g[i] * gamma[i] for i in range(n)]
